### ISSUE TITLE
Add (or improve) doctests and doccomments for all props, invs, thms, and utilities.

### ISSUE THREAD

**Issue by yirkajk** [math1um/objects-invariants-properties#546]
Note that inline `#` comments are not the same. We should have important information in doccomments, not doctests.

**Comment by nvcleemp** [comment]
We could also look at how the Sage function `help` is implemented, and then try to find a way to provide meaningful help messages for graphs in `gt.sage`

**Comment by nvcleemp** [comment]
After #548 we could also consider using [Sphinx](http://www.sphinx-doc.org) and https://readthedocs.org/.

**Comment by yirkajk** [comment]
Some suggestions for testing:

- Too many tests is better than too few. Code is cheap.
- Test for both positive and negative cases
- Test for boundary cases: The empty graph, disconnected graphs, and so on. In fact, these tests help us be precise with our definitions.
- And, since we're being bad and coding this after the fact, we can do white-box testing, i.e. we can make tests based on our current code. Ideally, we'd like our tests to "visit" every part of our code. So, if some `if` statement is only visited sometimes, make sure there's a test that does that!

**Comment by nvcleemp** [comment]
Maybe the necessity of doctests should also be made clear in the CONTRIBUTING.md file. Currently it only mentions that they are a good tool.

**Comment by yirkajk** [comment]
Also, standard guidelines for Python docstrings: https://www.python.org/dev/peps/pep-0257/

- A one line summary should appear on the same line as the `"""` or on the next line, without puncuation at the end. Then, a blank line. Then, any other comments.
- Doctests should appear at the same level of indentation as the rest of the comments.

And, comments don't have to be too extensive. The code should be self-documenting.
If the code isn't understandable, possibly due to short varialbe names or whatever, then we should improve the code.

**Comment by yirkajk** [comment]
@nvcleemp Should we be using `sage: foo(bar)` as doctests still? Or should we be doing something that's just Python, no sage?

**Comment by nvcleemp** [comment]
No, we should stick to using the Sage-type doctesting. Our code will always rely on Sage, and we will need Sage to run the doctests even when we switch to Python code. For Sage development people also use the Sage-type doctesting.

**Comment by nvcleemp** [comment]
We could also have a look at the [Sage guidelines](http://doc.sagemath.org/html/en/developer/coding_basics.html#documentation-strings)

If we also start using Sphinx to create a website with documentation, then at least things look similar to Sage. This should only improve the integration with Sage.
